jconstraints-cvc4/src/test/java/io/github/tudoaqua/jconstraints/cvc4/solver/UNSATCoreTest.java

```diff
@@ -0,0 +1,143 @@
+
+public class UNSATCoreTest extends AbstractCVC4Test {
+
+  @Test
+  public void example1Test() {
+    UNSATCoreSolver cvc4UnsatCore = (UNSATCoreSolver) cvc4;
+    cvc4UnsatCore.enableUnsatTracking();
+    SolverContext ctx = cvc4.createContext();
+
+    Variable p = Variable.create(BuiltinTypes.BOOL, "p");
+    Variable q = Variable.create(BuiltinTypes.BOOL, "q");
+    Variable r = Variable.create(BuiltinTypes.BOOL, "r");
+    Variable s = Variable.create(BuiltinTypes.BOOL, "s");
+
+    PropositionalCompound pc2 = PropositionalCompound.create(r, IMPLY, s);
+    PropositionalCompound pc3 =
+        PropositionalCompound.create(s, IMPLY, PropositionalCompound.create(q, EQUIV, r));
+    ctx.add(PropositionalCompound.create(p, OR, q));
+    ctx.add(pc2);
+    ctx.add(pc3);
+    ctx.add(PropositionalCompound.create(r, OR, p));
+    ctx.add(PropositionalCompound.create(r, OR, s));
+    ctx.add(Negation.create(PropositionalCompound.create(r, AND, q)));
+    ctx.add(Negation.create(PropositionalCompound.create(s, AND, p)));
+    assertEquals(Result.UNSAT, ctx.solve(null));
+    UNSATCoreSolver unsatCTX = (UNSATCoreSolver) ctx;
+
+    List<Expression> unsatCore = unsatCTX.getUnsatCore();
+    assertEquals(6, unsatCore.size());
+    assertTrue(unsatCore.contains(pc2));
+    assertFalse(unsatCore.contains(pc3));
+  }
+
+  @Test
+  public void example2Test() {
+    ConstraintSolver cvc4 = new CVC4Solver(new HashMap<>());
+    UNSATCoreSolver cvc4UnsatCore = (UNSATCoreSolver) cvc4;
+    cvc4UnsatCore.enableUnsatTracking();
+    SolverContext ctx = cvc4.createContext();
+    ctx.push();
+    ctx.pop();
+    ctx.push();
+
+    Variable p = Variable.create(BuiltinTypes.BOOL, "p");
+    Variable q = Variable.create(BuiltinTypes.BOOL, "q");
+    Variable r = Variable.create(BuiltinTypes.BOOL, "r");
+    Variable s = Variable.create(BuiltinTypes.BOOL, "s");
+
+    PropositionalCompound pc2 = PropositionalCompound.create(r, IMPLY, s);
+    PropositionalCompound pc3 =
+        PropositionalCompound.create(s, IMPLY, PropositionalCompound.create(q, EQUIV, r));
+    ctx.add(PropositionalCompound.create(p, OR, q));
+    ctx.add(pc2);
+    ctx.add(pc3);
+    ctx.add(PropositionalCompound.create(r, OR, p));
+    ctx.add(PropositionalCompound.create(r, OR, s));
+    ctx.add(Negation.create(PropositionalCompound.create(r, AND, q)));
+    ctx.add(Negation.create(PropositionalCompound.create(s, AND, p)));
+    assertEquals(Result.UNSAT, ctx.solve(null));
+    UNSATCoreSolver unsatCTX = (UNSATCoreSolver) ctx;
+
+    List<Expression> unsatCore = unsatCTX.getUnsatCore();
+    assertEquals(6, unsatCore.size());
+    assertTrue(unsatCore.contains(pc2));
+    assertFalse(unsatCore.contains(pc3));
+  }
+
+  @Test(
+      groups = {"manual"},
+      enabled = false)
+  public void example3Test() {
+    ConstraintSolver cvc4 = new CVC4SMTCMDSolver();
+    UNSATCoreSolver cvc4UnsatCore = (UNSATCoreSolver) cvc4;
+    cvc4UnsatCore.enableUnsatTracking();
```

Comment:
Upcasting requires downcasting in the next line
```suggestion
    CVC4SMTCMDSolver cvc4 = new CVC4SMTCMDSolver();
    cvc4.enableUnsatTracking();
```